File: project/jimmer-core/src/test/java/org/babyfish/jimmer/DataTest.java
package org.babyfish.jimmer;

import com.fasterxml.jackson.core.JsonProcessingException;
import org.babyfish.jimmer.model.Data;
import org.babyfish.jimmer.model.DataDraft;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

public class DataTest {

    @Test
    public void testSerialize() {
        Data data = DataDraft.$.produce(draft -> {
            draft.setList(Arrays.asList(1L, 2L));
            draft.setNestedList(
                    Arrays.asList(
                            Arrays.asList(3L, 4L),
                            Arrays.asList(5L, 6L)
                    )
            );
            draft.setArr(new long[] { 7L, 8L });
            draft.setNestedArr(
                    new long[][] {
                            { 9L, 10L },
                            { 11L, 12L }
                    }
            );
        });
        Assertions.assertEquals(
                "{\"list\":[1,2]," +
                        "\"nestedList\":[[3,4],[5,6]]," +
                        "\"arr\":[7,8]," +
                        "\"nestedArr\":[[9,10],[11,12]]}",
                data.toString()
        );
    }

    @Test
    public void testDeserialize() throws JsonProcessingException {
        String json = "{\"list\":[1,2]," +
                "\"nestedList\":[[3,4],[5,6]]," +
                "\"arr\":[7,8]," +
                "\"nestedArr\":[[9,10],[11,12]]}";
        Data data = ImmutableObjects.fromString(Data.class, json);
        Assertions.assertEquals(json, data.toString());
    }
}

File: project/jimmer-core-kotlin/src/test/kotlin/org/babyfish/jimmer/kt/DataTest.kt
package org.babyfish.jimmer.kt

import org.babyfish.jimmer.ImmutableObjects
import org.babyfish.jimmer.kt.model.Data
import org.babyfish.jimmer.kt.model.by
import kotlin.test.Test
import kotlin.test.expect

class DataTest {

    @Test
    fun testSerialize() {
        val data = new(Data::class).by {
            list = listOf(1L, 2L)
            nestedList = listOf(
                listOf(3L, 4L),
                listOf(5L, 6L)
            )
            arr = longArrayOf(7L, 8L)
            nestedArr = arrayOf(
                longArrayOf(9L, 10L),
                longArrayOf(11L, 2L)
            )
        }
        expect(
            """{"list":[1,2],"nestedList":[[3,4],[5,6]],"arr":[7,8],"nestedArr":[[9,10],[11,2]]}"""
        ) {
            data.toString()
        }
    }

    @Test
    fun testDeserialize() {
        val json = """{"list":[1,2],"nestedList":[[3,4],[5,6]],"arr":[7,8],"nestedArr":[[9,10],[11,2]]}"""
        val data = ImmutableObjects.fromString(Data::class.java, json)
        expect(json) {
            data.toString()
        }
    }
}